### ISSUE TITLE
Provide information about hard drive space and usage

### ISSUE THREAD

**Issue by olevski** [SwissDataScienceCenter/renku-notebooks#727]
This issue has remnants of a general discussion about how to communicate disk usage to users. Anything that is older than February 2022 can be ignored.

Some of the issues raised here are addressed by recent PRs such as: #923.

The resource usage as well as requests should be returned in the /servers/<name> and /servers endpoints.

The response should use the following schema:
```python
class ResourceUsage(Schema):
    cpu = fields.Str(required=False)
    memory = fields.Str(required=False)
    storage = fields.Str(required=False)

class ResourceRequests(ResourceUsage):
    gpu = fields.Str(required=False)

    @pre_load
    def resolve_gpu_fieldname(self, in_data, **kwargs):
        if "nvidia.com/gpu" in in_data.keys():
            in_data["gpu"] = in_data.pop("nvidia.com/gpu")
        return in_data

class Resources(Schema):  # Simply will replace UserPodResources
    requests = fields.Nested(ResourceRequests(), required=True)
    usage = fields.Nested(ResourceRequests(), required=False)
```

The most current data on resource usage from the session can be extracted from the JupyterServer resource manifest at `.status.mainPod.resourceUsage`. Here is an example of the format of the field:

```yaml
    resourceUsage:
      cpuMillicores: 5.633821999999999
      disk:
        availableBytes: 1002876928
        totalBytes: 1006526464
        usedBytes: 3649536
      memoryBytes: 80433152
```

What is currently being returned by the notebook service is what was requested when the session was launched. With this we will return what was requested AND what is being used.

Breaking API change requires a follow up from the UI.

**NOTE**: You cannot assume that `.status.mainPod` will be there all the time and same goes for any nested fields. The JupyterServer can exist in the cluster with just `status` being blank or definitely not containing even `mainPod` when sessions are starting`. Also if the metrics server is not enabled in the k8s cluster where amalthea is running then the cpu usage will be blank. It is also possible that due to problems with the image that is running the sessions also other parameters are missing.

**Comment by rokroskar** [comment]
Don't we need to inform the user _before_ the launch? And shouldn't this be the domain of the core service, since it's in the business of cloning repos and executing actions on them?

**Comment by olevski** [comment]
@rokroskar yes it would be ideal if we can let the user know that the selected hard drive space is not enough for their session before or immediatelly after they launch a session.

By putting this in the hands of the core service as you suggest it would mean that the UI calls the core service to validate whether the storage space selected by the user is enough for the commit they are launching a session from? If that is the case then yeah that makes sense for sure.

I dont really care that much which service takes care of this. I thought it was logical to put in the notebook service since the notebook service launches the sessions so it could potentially respond with an error when it receives a request to launch a session that will not have enough storage space.

**Comment by rokroskar** [comment]
It might be quickest to do it in the notebook service. We already have the necessary credentials to do the actual clone and since we just need a shallow clone it should be quite quick. @lorenzo-cavazzi probably has opinions about how the API for this should look...

**Comment by lorenzo-cavazzi** [comment]
This problem seems to be related to the storage provided by the k8s resources, are you sure it would be better addressed in renku-core than in renku-notebooks?

I think the initial idea from Tasko ( [here](https://github.com/SwissDataScienceCenter/renku-notebooks/pull/725#issuecomment-896792900) ) was reasonable. Of course, if it's easy to check beforehand that the storage is enough, that would be nice, though I guess that is not really needed most of the time.
Also, how should this check work? Do we consider only the repository files? Or do we check the lfs files? How can we know if the user needs all the lfs files at once if he didn't select the auto-fetch option?

**Comment by rokroskar** [comment]
yes exactly - that's why I initially thought such an endpoint would belong more in the core service. I agree with @lorenzo-cavazzi that since this is potentially an expensive operation we want to be careful not to do it unnecessarily and I think limiting the use only to the situation when the LFS autofetch checkbox is ticked would be sufficient.

**Comment by lorenzo-cavazzi** [comment]
I see one problem with this.
Currently, the UI doesn't know much about the notebooks options that are dynamically fetched when needed, and they are treated like meaningless variables.
Trying to codify in the `GET server_options` response the information that a specific API should be invoked when a certain option is ticked doesn't look that trivial since, in this case, it should be invoked also when other project-related parameters are changed (e.g. branch or commit).

An alternative would be to keep the current flow and perform the check as the first step in the `POST /servers` API. If the space won't be enough, we return an error and the UI shows it to the user.
This may not be as nice as having a real-time validation, but it would limit the invocations and it would mix better with the quick-start button logic we have recently added to the UI.

I guess in that case, renku-notebook may even try to start the pod with more space if that is possible. Or maybe this would make it even more confusing for the user :thinking:

**Comment by olevski** [comment]
After discussing things with @ableuler we can provide this information in the amalthea jupyterserver resource. The information will be periodically updated and saved in the k8s resource itself. Then the notebooks api can provide this info in the server objects in all the responses. Will touch base with @lorenzo-cavazzi on the format of this data.

**Comment by olevski** [comment]
@ableuler I wanted to outline here how I intend to tackle this because it depends on changes on the Amalthea and renku-notebooks side. Also on renku notebooks we need to enable the rpc server sidecar so that we can check how much free space there is.

Renku notebooks changes:
- Generate a random token used to authenticate with the rpc server
- Store said token in the secret that is part of the JupyterServer resource
- The rpc server has this secret injected in its container via an env variable
- When a request comes in to the rpc server then check if the request header contains this token, if it does and it matches then proceed normally, otherwise return an error
- Anyone who can access and parse the JupyterServer secret can use the rpc server
- Remaining questions:
  - How to open access to the rpc server? Presently we have a network policy (coming from Amalthea) which prevents ingress into any JupyterServer pod on any port other than 4180. However the rpc server needs another port. Therefore we can add a 2nd network policy through a path when a server is created in the notebook service to enable igress on port 4000 to the rcp server. What do you think about this?

Amalthea:
- Add a timer in Amalthea that will hit the rpc server and adjust the JupyterServer resource status with things like CPU, Disk usage, git status, etc.
- The question is how to make this flexible enough for more general usage and not just what we need. One thing that comes to mind is that the crd for JupyterServer contains a portion like this:
```yaml
liveStatusUpdates:
  - url: http://service-name:4000/jsonrpc
    type: POST
    payload:
      method: git/status
      jsonrpc: 2.0
      id: 0
    header: 
      AuthToken: 
        secretKeyRef: 
          name: jupyter-server-secret
          key: rcpServerAuthToken
    statusKey: liveStatus
```
- the timer in Amalthea then looks for this section in the server manifest and goes through all elements in the list, makes all the requests to get the new data and then updates the resource under each key specified in `statusKey`. The responses are not parsed but as long as the response is code 200 they are added on to the resource's status as text. We can have some limit on how long this text can be.

Now that I see all this written up like this perhaps it is maybe not a good idea to add this functionality in Amalthea. And we should depend on another service to query the rpc server and get the required information. I.e. when someone requests the list of servers from renku-notebooks then the notebook service does calls to the rpc server to get this information and insert it into the response.

**Comment by ableuler** [comment]
I agree with @rokroskar that the desired UX would be for the user to be _told in advance_ that the chosen combination of settings (branch, auto-fetch lfs, disk size) will not work. However, I think it's out of scope for the notebooks service to implement this functionality. On the other hand I think it is the responsibility of the notebook service to fail gracefully when running out of space and to inform the user accordingly. So I would opt for something along the following lines:

- We use the gitlab project statistics to warn a user whenever space _might_ get tight. So if the reported repo size ( + the LFS objects in case the auto fetch option is selected) is >~ half the available disk space, we show a banner telling the user that space might be an issue. I know that using the GitLab api for this is not ideal, but currently it is the best available information on wether a clone will work.

- If cloning the repo and checking out the desired branch fails because we're running out of disk space. I think the session launch should fail entirely and the user should be told, maybe while providing some suggestions with what to try (select bigger volume, uncheck the auto-lfs option).

From a technical perspective, running out of disk space during in the init container could lead to a failure with a predefined, non-0 exit code of that container. This could then be picked up by a specifically registered even handler listening to events on the main pod and updating the custom resource status accordingly. This would be contingent on the possibility of defining additional handlers for specific use cases.

**Comment by olevski** [comment]
@ableuler @ciyer are you ok with me making this into an epic with the following, more concrete subtasks:
- use the gitlab project statistics to warn a user whenever space might get tight
- if cloning step fails session launch should fail
- surface disk, cpu and memory usage info from Jupyterserver resources through the notebooks service

**Comment by olevski** [comment]
The following schema will be used to define snapshots of recent resource usage by the user sessions:

```python
class ResourceRequests(Schema):
    cpu = fields.Str(required=True)
    memory = fields.Str(required=True)
    storage = fields.Str(required=False)
    gpu = fields.Str(required=False)


class ResourceUsage(Schema):
    cpu = fields.Str(required=False)
    memory = fields.Str(required=False)
    storage = fields.Str(required=False)


class Resources(Schema):
    requests = fields.Nested(ResourceRequests(), required=True)
    usage = fields.Nested(ResourceRequests(), required=False)

```

**Comment by olevski** [comment]
We cannot reliably provide any warning about the resources required by a session before it is launched. Or whether the selected/requested resources will be enough. But providing this information may prove useful. At least in the case of disk usage.

The information that will be provided by the notebook service comes from amalthea. Amalthea periodically gets a snapshot of the resource usage and records it in the jupyterServer resource itself. The notebook service will simply relay this information to the ui.

**Comment by rokroskar** [comment]
I wonder if gathering information about dataset size could be some on-going background task that would additionally annotate the KG... Since we can't trust the stats that git/git-lfs provide and in any case we might have a more diverse set of storage backends in the future. Then at least for projects that use datasets the KG could give some info about the expected data volume.
